### PR TITLE
Gopher support (light version)

### PR DESCRIPTION
**Pull request by c0dev0id**
Hello,

I took the gopher pull request from @Ygrex (PR: https://github.com/luakit/luakit/pull/663) and and adapted it to my needs.

No offense to the author, but I did not like the inline iframe support. I changed the visual to match my favorite gopher client "sacc". I also changed the icons back to text and made sure that the alignment matches because many gopher sites use ASCII art and this would be ripped apart otherwise.

I also added a rudimentary stylesheet (which could be made themable or moved to rc.lua).

scr-2021-01-17_19-57-Iegh2yai

### CODE REVIEW

**Comment by c0dev0id** [comment]
@aidanholm what do you think about this version?
@Ygrex What about you? Most of the code is yours so I'd be fine if you rebase it and send it under your name. Credit where credit is due...

File: lib/gopher.lua [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+local error_page = require("error_page")
+local webview = require("webview")
+
+local _M = {}
+
+luakit.register_scheme("gopher")
+
+-- menu entry button's inner HTML representing the type of the item
+local function gophertype_to_icon(gophertype)
+    -- TODO unicode symbols look different
```
Comment (Ygrex) [inline lib/gopher.lua]: there is no unicode chars anymore, please remove the comment

File: lib/gopher.lua [hunk 2]
```diff
@@ -0,0 +1,476 @@
...
+end
+
+local error_page = require("error_page")
+local webview = require("webview")
+
+local _M = {}
+
+luakit.register_scheme("gopher")
+
+-- menu entry button's inner HTML representing the type of the item
```
Comment (Ygrex) [inline lib/gopher.lua]: it is no longer a button, just symbolical representation

File: lib/gopher.lua [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+    };
+    local line_num = 0
+    for line in data:gmatch("(.-)\r?\n") do
+        line_num = line_num + 1
+        local entry = parse_gopher_line(line, url)
+        if entry.item_type == "i" then
+            html[#html + 1] = ("    | %s"):format(entry.display_string)
+        else
+            local src = href_source(entry)
+            local button = gophertype_to_icon(entry.item_type) .. "+"
```
Comment (Ygrex) [inline lib/gopher.lua]: let's rename `button` to `icon` or anything, there are no buttons anymore

**Comment by Ygrex** [comment]
> @aidanholm what do you think about this version?
> @Ygrex What about you? Most of the code is yours so I'd be fine if you rebase it and send it under your name. Credit where credit is due...

Please proceed with this MR as it goes to avoid complexities with mutual rebases etc. Credits will not appear in Git log, but stay in the module header anyway, if anybody cares

Originally I wrote this module when Overbite stopped working on Firefox and I wanted to experiment with adding new schema to WebKit. That is why these iframes were there, it comes from Overbite. After trying this simplified version, I personally find it more attractive, iframes never looked as I wanted them to (wrong size, ratio, unable to fold back…). So I am actually going to reject my own PR in favor of this one.

**Comment by Ygrex** [comment]
one more suggestion for perfectionists is the commit description, it says `protocoll`

**Comment by c0dev0id** [comment]
I'm glad this is okay with you. I'll polish it further.

**Comment by c0dev0id** [comment]
I did some cleanup and added some support to set the fg, bg and link color via `theme.lua`. I also added the gopher protocol as comment to `rc.lua`.

Edit: regarding "protocoll" - yes I know. When we're done with this PR I'll do a sqash merge and correct the commit title.

**Review by Retro64XYZ — APPROVED** [review]
Very excited to see these changes. I have built a Gopher capable copy of Luakit and did some testing. Everything seemed to work.

OS - Manjaro Linux i3 Community Edition
Lua Version - 5.1

I started at gopher://gopher.club and moved between many different phlogs. Everything seemed to work and I had no issues. I then switched to the HiDPI mode and tested. Still all worked.

This is a fantastic addition and I am very excited about the potential for this to move into the normal build. Thank you for your effort and all you have done to add this. Great work!

Items needed to be installed included lua51-socket, nuspell, lua51-filesystem, and luarocks to successfully build and run everything. For anyone looking to test. All available in pacman.

Thanks!

**Comment by c0dev0id** [comment]
@Retro64XYZ Thanks for testing. Yes, luasocket is an additional dependency it brings in. It's only needed for those that will actually load the gopher module. I think we should ship it commented out, so users can load it via `userconf.lua`.

I'll add a comment about luasocket, so people know they need it when loading the module.

**Comment by c0dev0id** [comment]
@aidanholm, not sure if you're around. I'd merge this next week if nobody speaks up against it.

**Comment by aidanholm** [comment]
@c0dev0id sorry for the delay -- been very busy with work these few months. Feel free to merge anything you think looks good :) I can always make adjustments later if needed

This change specifically looks awesome! :tada:

**Comment by c0dev0id** [comment]
@aidanholm: Thanks, luakit is my favorite gopher browser now 🥰 I'll drive luakit forward  as time permits.  Feel free to step in any time.

**Pull request merged**